### PR TITLE
Added JWT authentication to the API

### PR DESCRIPTION
**Pull request by jclapis**
Based on @poupas's suggestions, this adds authentication to the daemon API via JWT. The daemon now requires standard JWT authorization from clients; the key passed in via `-k` must also be used to sign JWTs for any client requests.

### CODE REVIEW

File: shared/auth/auth-mgr.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+	"github.com/golang-jwt/jwt/v4"
+)
+
+const (
+	AuthorizationHeader string = "Authorization"
+	BearerPrefix        string = "Bearer "
+
+	// How long a JWT is valid for after sending a request
+	ValidTime time.Duration = time.Second * 5
```
Comment (huy-nodeset) [inline shared/auth/auth-mgr.go:19]: what is * 5 here?

**Review by huy-nodeset — APPROVED** [review]
1 question otherwise LGTM

File: shared/auth/auth-mgr.go (hunk 1 above)
Comment (jclapis) [reply shared/auth/auth-mgr.go:19]: It means 5 seconds, that's how you do const time durations. Basically when the daemons / CLI issue a request to another daemon, that request has to make it to the remote end within 5 seconds for it to be valid. Since everything's on the same machine it's basically immediate and this is kind of a formality anyway.

File: shared/auth/generate.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+)
+
+const (
+	// The default length of the API authorization key, in bytes
+	DefaultKeyLength int = 32
```
Comment (poupas) [inline shared/auth/generate.go:13]: I noticed you're using HMAC-SHA384 above. The [HMAC RFC](https://datatracker.ietf.org/doc/html/rfc2104) states that:

> The definition of HMAC requires a cryptographic hash function, which
   we denote by H, and a secret key K. We assume H to be a cryptographic
   hash function (...)   We denote by B the byte-length of such
   blocks (...) and by L the byte-length of hash outputs (L=16 for MD5, L=20 for
   SHA-1).

> The key for HMAC can be of any length (keys longer than B bytes are
   first hashed using H).  However, less than L bytes is strongly
   discouraged as it would decrease the security strength of the
   function.  Keys longer than L bytes are acceptable but the extra
   length would not significantly increase the function strength. (A
   longer key may be advisable if the randomness of the key is
   considered weak.)

So, tl;dr, you should bump the key size to 384/8 = 48 bytes.

Note: 256-bit key is fine. This comment is so you're protected against crypto-pedants.

Comment (jclapis) [reply shared/auth/generate.go]: Done - it now enforces that `GenerateAuthKeyIfNotPresent` will use *at least* 48 bytes. The actual key loader doesn't check the size (other than > 0) but we can enforce that too if it's prudent.

**Review by huy-nodeset — APPROVED** [review]
